2-Draft-Accepted/RFCNNNN-Support-Experimental-Features.md

```diff
@@ -0,0 +1,626 @@
+so that there is no duplication of the filtering logic for the `[Experimental()]` attribute.
+
+As for the type members with the `[Experimental()]` attributes,
+[`TypeDefiner`](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/parser/PSType.cs#L17) needs to be updated to process or skip a type member when processing a `TypeDefinitionAst` node,
+depending on whether the experimental feature is enabled and the `ExperimentAction` value.
+
+#### Exposed Public APIs
+
+An experimental feature may expose public C# APIs.
+Public C# APIs cannot be hidden.
+Even if we alter the `DotNetAdapter` to filter out those public APIs from the experimental feature in PowerShell,
+they will still be visible to applications that host PowerShell.
+Therefore, we will not do anything about the public APIs exposed from experimental features.
+
+### Experimental Feature Discovery
+
+A user should be able to discover the available experimental features in PowerShell,
+including the ones in the PowerShell engine as well as the ones from modules.
+
+#### PowerShell Engine Experimental Feature
+
+For experimental features in PowerShell engine,
+an internal static read-only collection `EngineExperimentalFeatures` will be used to track all available engine experimental features.
+The items in the collection are of the type `ExperimentalFeature`, whose definition looks as follows:
+
+```c#
+public class ExperimentalFeature
+{
+    // Feature name
+    public string Name { get; }
+    // Feature description
+    public string Description { get; }
+    // Feature source (e.g. PSEngine, or module name)
+    public string Source { get; }
+    // Feature is enabled or not
+    public bool IsEnabled { get; }
+}
+
+internal static readonly ReadOnlyCollection<ExperimentalFeature> EngineExperimentalFeatures;
+```
+
+When adding a new experimental feature, an instance representing the feature needs to be added to the collection for it to be discoverable to users.
+The `Source` for engine experimental features should be specified as `PSEngine`.
+
+#### Module Experimental Feature
+
+For experimental features in modules,
+metadata about an experimental feature should be kept in the module manifest.
+We use the `PrivateData` section of a module manifest to expose the experimental features from the module.
+Here is an example:
+
+```powershell
+PrivateData = @{
+    ExperimentalFeatures = @(
+        @{Name = "PSWebCmdletV2", Description = "Rewrite the web cmdlets for better performance"}
+        @{Name = "PSRestCmdletV2", Description = "Rewrite the REST API cmdlets for better performance"}
+    )
+}
+```
+
+PowerShell module components, such as `Import-Module` and module analysis,
+will be updated to incorporate this metadata to the resulted `PSModuleInfo` object.
+
+#### Get-ExperimentalFeature
+
+A new cmdlet `Get-ExperimentalFeature` will be added to return all available experimental features of a PowerShell session.
```

Comment:
What do you mean by 'available'?  Does this include 'hidden' experimental features?  Should there be an option to show 'hidden' experimental features?

Reply:
Here "available" means all experimental features, enabled or not enabled. The returned `ExperimentalFeature` type has a `IsEnabled` property, so you will know which ones are enabled/not enabled.

Reply:
@PaulHigin - experimental implies that people should experiment with these features hence the need for discoverability. What would be the purpose of a "hidden" experimental feature? (And since it's open source, nothing is really hidden)

Reply:
@BrucePay: There is no concept of a hidden experimental feature.  In the referenced statement, available could be removed.